### DISCUSSION TITLE
Howto set RadioSet selection programmatically

### DISCUSSION THREAD

**Discussion by UlrichGoebel** [Textualize/textual#4572] [Q&A]
Hi,

I try to set the actual choice of a `RadioSet` programmatically an struggle in understanding the `RadioSet`. To demonstrate my struggeling let's see the following code:

```python
from textual.app import App
from textual.widgets import Button, Input, Label, RadioButton, RadioSet

class Main(App):
  
    def compose(self):
        with RadioSet():
            yield RadioButton('One',   id='one',    value=True)
            yield RadioButton('Two',   id='two')
            yield RadioButton('Three', id='three')
        yield Button('Action', id='action')
        yield Button('Show', id='show')
        yield Label('Actual Choice:')
        yield Input()
    
    def on_button_pressed(self, event):
        if event.button.id == 'action':
            self.query_one(RadioSet).action_next_button()   # may be commanded out
            self.query_one(RadioSet).action_toggle_button() # may be commanded out
        # In any case show which RadioButton is active
        self.query_one(Input).value = self.query_one(RadioSet).pressed_button.id
    
if __name__ == "__main__":
    app = Main()
    app.run()
```

It shows a `RadioSet` with three choices, then two `Buttons`, one for showing the actual choice, the other to trigger some action(s) on the `RadioSet`.

The `on_button_pressed` handler should start the action or even two actions, if the `Action` button is pressed. Independend which button was pressed it should update the value of the `Input` to show the actual/new choice.

That is what I expect. But...

1. Clicking the `Action` Button changes the choice (this can be seen in the terminal) but in the `Input` it shows the _last_ choice, not the _actual_. Why?
2. Clicking the `Show` Button shows the actual choice as expected.
3. Commanding out the `action_toggle_button`, so just `action_next_button` is done: Well, the choice doesn' change. That's o.k. But what does it actually?
4. Commanding out the `action_next_button`, so just `action_toggle_button` should be done: Clicking the `Action` doesn't do anything but showing the actual choice. So it seems that it can`t toggle the choice if it is alredy activated - which obviously makes sense in a `RadioSet` context because it can't know which `RadioButton` has to be the new choice. Does that mean that _toggle_ in fact means _set_?
5. There seems to be a difference between _selecting_ a `RadioButton` and _choose_ it.
6. Is there a possibility to get the _chosen_ `RadioButton` even if it is not the _pressed_ one? Then I could step through the `RadioButton` using `action_next_button` until I found the right one which can then be toggled by `action_toggle_button`, which in fact means it would be the new choice.

Can someone help me to understand what's going on here?

Best regards
Ulrich

**Comment by TomJGooding** [comment]
This looks like a possible bug, rather than something you're doing wrong.

I don't yet fully understand the workings of the `RadioSet`, but my quick test also doesn't seem to work as expected. This might need some input from a maintainer...

```python
from textual.app import App, ComposeResult
from textual.widgets import Footer, RadioButton, RadioSet


class ExampleApp(App):
    BINDINGS = [("n", "toggle_next_button", "Toggle Next Buttton")]

    def compose(self) -> ComposeResult:
        with RadioSet():
            yield RadioButton("One", id="one", value=True)
            yield RadioButton("Two", id="two")
            yield RadioButton("Three", id="three")

        yield Footer()

    def action_toggle_next_button(self) -> None:
        radio_set = self.query_one(RadioSet)
        assert radio_set._selected == 0

        radio_set.action_next_button()
        assert radio_set._selected == 1

        radio_set.action_toggle_button()
        assert radio_set._selected == 1
        assert radio_set.pressed_button is not None
        # The test below will fail, despite the toggle action seemingly using
        # `_selected` attribute as an index?
        assert radio_set.pressed_index == 1


if __name__ == "__main__":
    app = ExampleApp()
    app.run()

```
